Add support for LTR390

sjtrny opened the pull request:
## Description:

Adds support for the LTR390 sensor, which is a UV and ambient light sensor.

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** https://github.com/esphome/esphome-docs/pull/979

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

sjtrny committed:
Add support for ltr390

sjtrny committed:
Fix linting errors

sjtrny committed:
Fix more linting errors

sjtrny committed:
Linting fixes continued

sjtrny committed:
Linting forever

sjtrny committed:
Another one

sjtrny committed:
Fix regression and linting

sjtrny committed:
Fix narrowing conversion

sjtrny committed:
Add test and bugfix

sjtrny committed:
Add codeowners

sjtrny committed:
Update CODEOWNERS

sjtrny committed:
Update sensor defs

sjtrny committed:
Reformatted with black

sjtrny committed:
Fixed device class import

sjtrny committed:
Update CODEOWNERS

sjtrny committed:
Update CODEOWNERS

jesserockz commented on tests/test1.yaml:
```diff
@@ -565,6 +565,15 @@ sensor:
     humidity:
       name: 'Living Room Humidity 6'
     update_interval: 15s
+  - platform: ltr390
+    uv:
+      name: "LTR390 UV"
+    uvi:
+      name: "LTR390 UVI"
+    light:
+      name: "LTR390 Light"
+    als:
+      name: "LTR390 ALS"
```
Can you please add **all** config options to the test file? This is so someone doesn't change the config later on and makes breaking changes. (Easier to notice when the test files are changed)

sjtrny replied:
I've added all config options for the sensor now.

sjtrny committed:
Adding all config options

As requested https://github.com/esphome/esphome/pull/1505#discussion_r597326897

sjtrny committed:
Moving test to different config file

test1.yml runs out of memory

OttoWinter committed:
Merge remote-tracking branch 'origin/dev' into ltr390

OttoWinter commented on esphome/components/ltr390/sensor.py:
```diff
@@ -0,0 +1,96 @@
+    "LTR390Component", cg.PollingComponent, i2c.I2CDevice
+)
+
+CONF_ALS = "als"
```
We should use long descriptive names here like `ambient_light` and `window_correction_factor`

OttoWinter commented on esphome/components/ltr390/ltr390.cpp:
```diff
@@ -0,0 +1,204 @@
+void LTR390Component::setup() {
+  ESP_LOGCONFIG(TAG, "Setting up ltr390...");
+
+  this->ctrl_reg_ = new i2c::I2CRegister(this, LTR390_MAIN_CTRL);
```
Hmm, the I2CRegister interface isn't really meant to be used this way (I'm working on some improvements, and those changes would break this).

I'll fix the constructor visibility, but this needs to be converted to use `this->reg(...)`

sjtrny replied:
Looks like you ended up doing it all here. Marking as resolved.

OttoWinter commented on esphome/components/ltr390/ltr390.cpp:
```diff
@@ -0,0 +1,204 @@
+  while (!this->new_data_available_()) {
+    ESP_LOGD(TAG, "WAITING FOR DATA");
+    delay(2);
+  }
```
If no data is ever available this will block forever. This needs some logic to abort after ~100ms or so

sjtrny replied:
Thanks for fixing. Marking as resolved.

OttoWinter commented on esphome/components/ltr390/ltr390.cpp:
```diff
@@ -0,0 +1,204 @@
+void LTR390Component::update() {
+  if (!this->reading_) {
+    this->reading_ = true;
+    this->read_mode_(0);
```
If none of the sensors are defined (which is a valid YAML config), then `mode_funcs_` will be empty and the read of `mode_funcs_` will fail.

sjtrny replied:
I just tried a config with empty sensors but got the following error

> Must contain at least one of light, ambient_light, uv_index, uv.

However I have put a check into the if condition to make sure there are some modes that need reading.

OttoWinter commented on esphome/components/ltr390/ltr390.h:
```diff
@@ -0,0 +1,114 @@
+
+  void read_mode_(int mode_index);
+
+  std::atomic<bool> reading_;
```
Unfortunately we can't really use atomics (at least on ESP8266 I think many of the threading related includes are not available)

sjtrny replied:
Is there a recommended way to prevent race conditions? I was running into problems when using a high update rate but long measurement time on the sensor.

OttoWinter replied:
ESPHome is single threaded (because on ESP8266 there's no threading support in the cpp stdlib). So there are no concurrent accesses to the same memory location, thus the atomic doesn't do anything here.

sjtrny replied:
Got it. You've fixed it so marking as resolved.

OttoWinter committed:
Update according to comments

OttoWinter reviewed (commented):
Thanks for this contribution! Sorry for the long review delay.

I just took a look at this PR, and found some minor things to address - I figured it's probably easier if I just push the changes to make it quicker.

Can you check if the component works with those changes? Thanks!

sjtrny committed:
Add safety clause to reading modes

sjtrny commented:
@OttoWinter I've tested changes locally and everything seems ok!

sjtrny committed:
Fix clang-tidy complaint

oxan committed:
Revert change to i2c component

oxan commented:
> @OttoWinter I've tested changes locally and everything seems ok!

Okay, it seems we can merge this then :) I've removed one change to the I2C component that, while it seems correct, probably ended up in this branch by accident.

jesserockz commented:
> Okay, it seems we can merge this then :) I've removed one change to the I2C component that, while it seems correct, probably ended up in this branch by accident.

Otto pushed that change into this PR (https://github.com/esphome/esphome/pull/1505/commits/a8a14c761b0d8da5705f60f54ed37515b4f57bd9)

oxan committed:
Merge branch 'dev' into ltr390

oxan committed:
Fix for changes in dev

oxan committed:
Revert "Revert change to i2c component"

This reverts commit 2810df59e9c05311df6d32149ed79a393676503b.

oxan commented:
> Otto pushed that change into this PR ([a8a14c7](https://github.com/esphome/esphome/commit/a8a14c761b0d8da5705f60f54ed37515b4f57bd9))

Ah, apparently I read over the part where he said he'd change the constructor visibility. Added it back in (though I do think changes like that are better done in separate PR's).
